packages/venia-concept/src/components/ResponsiveImage/ResponsiveImage.js

```diff
@@ -0,0 +1,166 @@
+/**
+ * Autogenerate an image tag with `sizes` and `srcSet` attributes, using simple
+ * media conditions, a source URL, and an array of pixel widths for resized
+ * images to create. Renders the image resize URLs expected by venia-upward.
+ *
+ * Optionally, takes a render callback so you can use other DOM structures
+ * besides literal img tags to display responsive images. The img tag is
+ * required, however.
+ */
+
+import React, { Component } from 'react';
+import { arrayOf, func, number, oneOf, string } from 'prop-types';
+import makeMediaUrl from 'src/util/makeMediaUrl';
+
+class ResponsiveImage extends Component {
+    static propTypes = {
+        alt: string.isRequired,
+        className: string,
+        render: func,
+        sizes: string.isRequired,
+        src: string.isRequired,
+        type: oneOf(['product', 'category', 'other']).isRequired,
+        widthOptions: arrayOf(number).isRequired
+    };
+
+    /**
+     * ## Updating the Background Image
+     *
+     * TODO: Image elements use srcsets to change their source file according to
+     * media conditions, but the equivalent CSS function `image-set` is not well
+     * supported yet.
+     *
+     * Users of ResponsiveImage may pass a render prop to render custom DOM
+     * using the responsive image element and its current source value.
+     *
+     * This render function probably needs the image URLs generated by
+     * ResponsiveImage (otherwise there would be no point in a render prop), so
+     * we have to keep track of them as they change.
+     *
+     * First we maintain a per-instance array of DOM elements awaiting update.
+     */
+
+    toSetStyle = [];
+
+    /**
+     * We will pass a reference to this `setToUrl` method to the render
+     * function. It takes one argument: either a string for a style attribute to
+     * set, or a function to run arbitrary code on the ref.
+     */
+
+    setToUrl = propToSet => {
+        const typeOfArg = typeof propToSet;
+        const ref = React.createRef();
+        /**
+         * Updating the `style` property with a CSS variable does not trigger a
+         * repaint, so we have to modify the string in the style attribute.
+         */
+        let callback;
+        if (typeOfArg === 'function') {
+            callback = propToSet;
+        } else if (typeOfArg === 'string') {
+            const styleRegexp = new RegExp(`${propToSet}:\\s*[^;]+(;?)`);
+            callback = (element, currentSrc) => {
+                const declaration = `${propToSet}: url("${currentSrc}")`;
```

Comment:
Would be good to investigate some alternatives here that don't (unless I'm missing something) result in a minimum of two requests per image: one for the low-res preview set here, one for the actual image. Maybe this could create an inline preview, i.e. super low-res then CSS blurred or just a background colour?
![image](https://user-images.githubusercontent.com/108052/51195539-2ba01680-18e5-11e9-8380-8ea6d2e702a3.png)

Reply:
Actually it's both better and worse: currently, UPWARD cannot rewrite URLs while proxying, so instead we're using 301 redirects--one for each image. They are to the same host, at least, but that's definitely two TCP handshakes. It's a tech debt item.